offchain/eth-input-reader/src/setup.rs

```diff
@@ -1,88 +1,116 @@
+    let http = Http::new(
+        Url::parse(&config.bh_config.http_endpoint).context(ParseSnafu)?,
+    );
+
+    let retry_client =
+        RetryClient::new(http, Box::new(HttpRateLimitRetryPolicy), 10, 1000);
+
+    let provider = Provider::new(retry_client);
 
-    Ok(GrpcStateFoldClient::new_from_channel(channel, config))
+    Ok(Arc::new(provider))
 }
 
-pub async fn create_block_subscription(
-    client: &impl BlockServer,
-    confirmations: usize,
+pub async fn create_block_subscriber(
+    config: &EthInputReaderConfig,
+    provider: Arc<InputProvider>,
+) -> Result<Arc<BlockSubscriber<InputProvider>>, EthInputReaderError> {
+    Ok(Arc::new(
+        BlockSubscriber::start(
+            Arc::clone(&provider),
+            config.bh_config.ws_endpoint.to_owned(),
+            config.bh_config.block_timeout,
+            config.bh_config.max_depth,
+        )
+        .await
+        .context(BlockArchiveSnafu)?,
+    ))
+}
+
+pub async fn create_subscription(
+    block_subscriber: Arc<BlockSubscriber<InputProvider>>,
+    subscription_depth: usize,
 ) -> Result<
-    impl Stream<Item = Result<BlockStreamItem, StateServerError>>
-        + Send
+    impl Stream<Item = Result<BlockStreamItem, SubscriptionError<InputProvider>>>
         + std::marker::Unpin,
     EthInputReaderError,
 > {
-    let s = client
-        .subscribe_blocks(confirmations)
+    Ok(block_subscriber
+        .subscribe_new_blocks_at_depth(subscription_depth)
         .await
-        .context(StateServerSnafu)?;
-
-    let s = {
-        use futures::StreamExt;
-        s.ready_chunks(BUFFER_LEN)
-    };
+        .context(BlockArchiveSnafu)?)
+}
 
-    let s = s.filter_map(
-        |mut x| {
-            if x.len() == BUFFER_LEN {
-                None
-            } else {
-                x.pop()
-            }
-        },
+pub async fn create_env(
+    config: &EthInputReaderConfig,
+    provider: Arc<InputProvider>,
+    block_archive: Arc<BlockArchive<InputProvider>>,
+) -> Result<
+    Arc<StateFoldEnvironment<InputProvider, Mutex<UserData>>>,
+    EthInputReaderError,
+> {
+    let env = StateFoldEnvironment::new(
+        provider,
+        Some(block_archive),
+        config.sf_config.safety_margin,
+        config.sf_config.genesis_block,
+        config.sf_config.query_limit_error_codes.clone(),
+        config.sf_config.concurrent_events_fetch,
+        MAXIMUM_EVENTS_PER_RESPONSE,
+        Mutex::new(UserData::default()),
     );
 
-    Ok(s)
+    Ok(Arc::new(env))
 }
 
 pub async fn create_context(
     config: &EthInputReaderConfig,
-    block_server: &impl BlockServer,
+    block_subscriber: Arc<BlockSubscriber<InputProvider>>,
     broker: &impl BrokerStatus,
     dapp_metadata: DAppMetadata,
     metrics: EthInputReaderMetrics,
 ) -> Result<Context, EthInputReaderError> {
-    let genesis_timestamp: u64 = block_server
-        .query_block(config.dapp_deployment.deploy_block_hash)
+    let genesis_timestamp: u64 = block_subscriber
```

Comment:
I don't think this makes sense. I think you should use the provider instead.

Reply:
How do I use the provider to get the genesis timestamp of a block?